File: mclimate/src/main/java/com/mordrum/mclimate/common/Season.java
package com.mordrum.mclimate.common;

import java.awt.*;
import java.util.GregorianCalendar;

//TODO hook into BlockColors and inject Biomes O' Plenty leaves
public enum Season {
    SPRING(0x000000, 0.0, 0x000000, 0.0), SUMMER(0x00FF00, 0.3, 0x00FF00, 0.3), AUTUMN(0x000000, 0.0, 0xFF6A00, 0.7), WINTER(0x999988, 0.9, 0x999988, 1.0);

    private final Color foliageColor;
    private final double foliageMultiplier;
    private final Color grassColor;
    private final double grassMultiplier;

    Season(int foliageColor, double foliageMultiplier, int grassColor, double grassMultiplier) {
        this.foliageColor = new Color(foliageColor);
        this.foliageMultiplier = foliageMultiplier;
        this.grassColor = new Color(grassColor);
        this.grassMultiplier = grassMultiplier;
    }

    public static Season getCurrentSeason() {
        return getSeasonForTimestamp(System.currentTimeMillis());
    }

    public static Season getSeasonForCalendar(GregorianCalendar calendar) {
        return getSeasonForTimestamp(calendar.getTimeInMillis());
    }

    private static Season getSeasonForTimestamp(long timestamp) {
        // Convert to seconds
        timestamp /= 1000;
        // -4 hour offset
        timestamp -= 4 * 60 * 60;

        int floor = (int) Math.floor(timestamp / 86400);
        int season = floor % 4;
        return Season.values()[season];
    }

    public int getFoliageColor(int originalColor) {
        return mixColors(new Color(originalColor), this.foliageColor, this.foliageMultiplier);
    }

    public int getGrassColor(int originalColor) {
        return mixColors(new Color(originalColor), this.grassColor, this.grassMultiplier);
    }

    private int mixColors(Color baseColor, Color colorToMix, double multiplier) {
        int red = (int) (baseColor.getRed() * (1F - multiplier) + colorToMix.getRed() * multiplier);
        int green = (int) (baseColor.getGreen() * (1F - multiplier) + colorToMix.getGreen() * multiplier);
        int blue = (int) (baseColor.getBlue() * (1F - multiplier) + colorToMix.getBlue() * multiplier);

        return new Color(red, green, blue).getRGB();
    }

    public float getModifiedTemperature(float temperature) {
        if (this == SUMMER) return temperature * 2;
        else if (this == WINTER) {
            if (temperature <= 0.8f) return 0.1f;
            else return temperature / 2;
        } else return temperature;
    }
}

File: mclimate/src/main/java/com/mordrum/mclimate/common/Season.kt
package com.mordrum.mclimate.common

import java.awt.*
import java.util.GregorianCalendar

//TODO hook into BlockColors and inject Biomes O' Plenty leaves
enum class Season(foliageColor: Int, private val foliageMultiplier: Double, grassColor: Int, private val grassMultiplier: Double) {
    SPRING(0x000000, 0.0, 0x000000, 0.0), SUMMER(0x00FF00, 0.3, 0x00FF00, 0.3), AUTUMN(0x000000, 0.0, 0xFF6A00, 0.7), WINTER(0x999988, 0.9, 0x999988, 1.0);

    private val foliageColor: Color = Color(foliageColor)
    private val grassColor: Color = Color(grassColor)

    fun getFoliageColor(originalColor: Int): Int {
        return mixColors(Color(originalColor), this.foliageColor, this.foliageMultiplier)
    }

    fun getGrassColor(originalColor: Int): Int {
        return mixColors(Color(originalColor), this.grassColor, this.grassMultiplier)
    }

    private fun mixColors(baseColor: Color, colorToMix: Color, multiplier: Double): Int {
        val red = (baseColor.red * (1f - multiplier) + colorToMix.red * multiplier).toInt()
        val green = (baseColor.green * (1f - multiplier) + colorToMix.green * multiplier).toInt()
        val blue = (baseColor.blue * (1f - multiplier) + colorToMix.blue * multiplier).toInt()

        return Color(red, green, blue).rgb
    }

    fun getModifiedTemperature(temperature: Float): Float {
        if (this == SUMMER)
            return temperature * 2
        else if (this == WINTER) {
            if (temperature <= 0.8f)
                return 0.1f
            else
                return temperature / 2
        } else
            return temperature
    }

    companion object {
        val currentSeason: Season
            get() = getSeasonForTimestamp(System.currentTimeMillis())

        fun getSeasonForCalendar(calendar: GregorianCalendar): Season {
            return getSeasonForTimestamp(calendar.timeInMillis)
        }

        private fun getSeasonForTimestamp(timestamp: Long): Season {
            var mutableTimestamp = timestamp
            // Convert to seconds
            mutableTimestamp /= 1000
            // -4 hour offset
            mutableTimestamp -= (4 * 60 * 60).toLong()

            val floor = Math.floor((mutableTimestamp / 86400).toDouble()).toInt()
            val season = floor % 4
            return Season.values()[season]
        }
    }
}
